File: src/Shared/PathUtil.m
//
//  PathUtil.m
//  OpenStreetMap
//
//  Created by Bryce Cogswell on 1/24/13.
//  Copyright (c) 2013 Bryce Cogswell. All rights reserved.
//

#import "PathUtil.h"


static void InvokeBlockAlongPathCallback2( void * info, const CGPathElement * element )
{
	ApplyPathCallback block = (__bridge ApplyPathCallback)info;
	block( element->type, element->points );
}

void CGPathApplyBlockEx( CGPathRef path, ApplyPathCallback block )
{
	CGPathApply(path, (__bridge void *)block, InvokeBlockAlongPathCallback2);
}

NSInteger CGPathPointCount( CGPathRef path )
{
	__block NSInteger count = 0;
	CGPathApplyBlockEx( path, ^(CGPathElementType type, CGPoint *points) {
		++count;
	});
	return count;
}

NSInteger CGPathGetPoints( CGPathRef path, CGPoint pointList[] )
{
	__block NSInteger index = 0;
	CGPathApplyBlockEx( path, ^(CGPathElementType type, CGPoint *points) {
		switch ( type ) {
			case kCGPathElementMoveToPoint:
			case kCGPathElementAddLineToPoint:
				pointList[ index++ ] = points[0];
				break;
			case kCGPathElementCloseSubpath:
				pointList[ index++ ] = pointList[0];
				break;
			default:
				break;
		}
	});
	return index;
}

void CGPathDump( CGPathRef path )
{
	CGPathApplyBlockEx( path, ^(CGPathElementType type, CGPoint *points) {
		NSLog(@"%f,%f", points->x, points->y );
	});
}

void InvokeBlockAlongPath( CGPathRef path, double initialOffset, double interval, void(^callback)(CGPoint pt, CGPoint direction) )
{
	__block CGFloat offset = initialOffset;
	__block CGPoint previous;

	void(^block)(CGPathElementType type, CGPoint * points) = ^(CGPathElementType type, CGPoint * points){

		switch ( type ) {

			case kCGPathElementMoveToPoint:
				previous = points[0];
				break;

			case kCGPathElementAddLineToPoint:
			{
				CGPoint nextPt = points[0];
				double dx = nextPt.x - previous.x;
				double dy = nextPt.y - previous.y;
				double len = sqrt( dx*dx + dy*dy );
				dx /= len;
				dy /= len;

				while ( offset < len ) {
					// found it
					CGPoint pos = { previous.x + offset * dx, previous.y + offset * dy };
					CGPoint dir = { dx, dy };
					callback( pos, dir );
					offset += interval;
				}
				offset -= len;
				previous = nextPt;
			}
				break;

			case kCGPathElementAddQuadCurveToPoint:
			case kCGPathElementAddCurveToPoint:
			case kCGPathElementCloseSubpath:
				assert(NO);
				break;
		}
	};
	CGPathApplyBlockEx( path, block );
}

void PathPositionAndAngleForOffset( CGPathRef path, double startOffset, double baselineOffsetDistance, CGPoint * pPos, CGFloat * pAngle, CGFloat * pLength )
{
	__block BOOL	reachedOffset = NO;
	__block BOOL	quit = NO;
	__block CGPoint	previous = { 0 };
	__block CGFloat	offset = startOffset;

	CGPathApplyBlockEx( path, ^(CGPathElementType type, CGPoint * points) {
		if ( quit )
			return;
		switch ( type ) {
			case kCGPathElementMoveToPoint:
				previous = points[0];
				break;
			case kCGPathElementAddLineToPoint:
			{
				CGPoint pt = points[0];
				CGFloat dx = pt.x - previous.x;
				CGFloat dy = pt.y - previous.y;
				CGFloat len = hypot(dx,dy);
				dx /= len;
				dy /= len;
				CGFloat a = atan2f(dy,dx);

				// shift text off baseline
				CGPoint baselineOffset = { dy * baselineOffsetDistance, -dx * baselineOffsetDistance };

				if ( !reachedOffset ) {
					// always set position/angle because if we fall off the end we need it set
					pPos->x = previous.x + offset * dx + baselineOffset.x;
					pPos->y = previous.y + offset * dy + baselineOffset.y;
					*pAngle = a;
					*pLength = len - offset;
				} else {
					if ( fabs(a - *pAngle) < M_PI/40 ) {
						// continuation of previous
						*pLength = len - offset;
					} else {
						quit = YES;
					}
				}

				if ( offset < len ) {
					// found it
					reachedOffset = YES;
				}
				offset -= len;
				previous = pt;
			}
				break;
			case kCGPathElementAddQuadCurveToPoint:
			case kCGPathElementAddCurveToPoint:
			case kCGPathElementCloseSubpath:
				assert(NO);
				break;
		}
	});
}

// reverse path
CGMutablePathRef PathReversed( CGPathRef path )
{
	NSMutableArray * a = [NSMutableArray new];
	CGPathApplyBlockEx( path, ^(CGPathElementType type, CGPoint * points){
		if ( type == kCGPathElementMoveToPoint || type == kCGPathElementAddLineToPoint ) {
			CGPoint cgPoint = points[0];
			OSMPoint pt = { cgPoint.x, cgPoint.y };
			OSMPointBoxed * boxed = [OSMPointBoxed pointWithPoint:pt];
			[a addObject:boxed];
		}
	});
	CGMutablePathRef newPath = CGPathCreateMutable();
	__block BOOL first = YES;
	[a enumerateObjectsWithOptions:NSEnumerationReverse usingBlock:^(OSMPointBoxed * pt, NSUInteger idx, BOOL *stop) {
		if ( first ) {
			first = NO;
			CGPathMoveToPoint( newPath, NULL, pt.point.x, pt.point.y );
		} else {
			CGPathAddLineToPoint( newPath, NULL, pt.point.x, pt.point.y );
		}
	}];
	return newPath;
}



static CGPoint * DouglasPeuckerCore( CGPoint points[], NSInteger first, NSInteger last, double epsilon, CGPoint * result )
{
	// Find the point with the maximum distance
	double dmax = 0.0;
	NSInteger index = 0;
	OSMPoint end1 = OSMPointFromCGPoint( points[first] );
	OSMPoint end2 = OSMPointFromCGPoint( points[last] );
	for ( NSInteger i = first+1; i < last; ++i ) {
		OSMPoint p = OSMPointFromCGPoint( points[i] );
		CGFloat d = DistanceFromPointToLineSegment( p, end1, end2 );
		if ( d > dmax ) {
			index = i;
			dmax = d;
		}
	}
	// If max distance is greater than epsilon, recursively simplify
	if ( dmax > epsilon ) {
		// Recursive call
		result = DouglasPeuckerCore( points, first, index, epsilon, result );
		result = DouglasPeuckerCore( points, index, last, epsilon, result-1 );
	} else {
		*result++ = CGPointFromOSMPoint( end1 );
		*result++ = CGPointFromOSMPoint( end2 );
	}
	return result;
}

CGMutablePathRef PathWithReducePoints( CGPathRef path, double epsilon ) CF_RETURNS_RETAINED
{
	NSInteger count = CGPathPointCount( path );
	if ( count < 3 )
		return CGPathCreateMutableCopy( path );
	CGPoint * points = (CGPoint *)malloc( count * sizeof(points[0]));
	CGPoint * result = (CGPoint *)malloc( count * sizeof(result[0]));
	CGPathGetPoints( path, points );
	CGPoint * resultLast = DouglasPeuckerCore( points, 0, count-1, epsilon, result );
	NSInteger resultCount = resultLast - result;
	CGMutablePathRef newPath = CGPathCreateMutable();
	CGPathAddLines( newPath, NULL, result, resultCount );
	free( points );
	free( result );
	return newPath;
}

File: src/Shared/PathUtil.swift
//
//  PathUtil.swift
//  OpenStreetMap
//
//  Created by Bryce Cogswell on 1/24/13.
//  Copyright (c) 2013 Bryce Cogswell. All rights reserved.
//

//#define OpenStreetMap_PathUtil_h
public typealias ApplyPathCallback = (CGPathElementType, CGPoint) -> Void

extension CGPath {
	func apply(action: (CGPathElement)->Void) {
		var action = action
		apply(info: &action, function: { (type,element) in
			let block = type!.bindMemory(to: ((CGPathElement)->()).self, capacity: 1).pointee
			block( element.pointee )
		})
	}
}

func CGPathPointCount(_ path: CGPath) -> Int {
    var count = 0
	path.apply(action: { _ in count += 1 })
    return count
}

func CGPathGetPoints(_ path: CGPath ) -> [CGPoint] {
	let count = CGPathPointCount( path )
	var list = [CGPoint]()
	list.reserveCapacity(count)
	path.apply(action: { element in
		switch element.type {
			case .moveToPoint, .addLineToPoint:
				list.append( element.points[0] )
			case .closeSubpath:
				list.append( list[0] )
			default:
				break
		}
	})
	return list
}

func CGPathDump(_ path: CGPath) {
	path.apply(action: { element in
		let point = element.points[0]
        print("\(point.x),\(point.y)")
    })
}

func InvokeBlockAlongPath(_ path: CGPath, _ initialOffset: Double, _ interval: Double, _ callback: (_ pt: CGPoint, _ direction: CGPoint) -> Void) {
    var offset = initialOffset
    var previous = CGPoint()

	path.apply(action: { element in
		switch element.type {
            case .moveToPoint:
				previous = element.points[0]
            case .addLineToPoint:
                let nextPt = element.points[0]
                var dx = Double((nextPt.x) - previous.x)
                var dy = Double((nextPt.y) - previous.y)
                let len = sqrt(dx * dx + dy * dy)
                dx /= len
                dy /= len

                while offset < len {
                    // found it
                    let pos = CGPoint(x: Double(previous.x) + Double(offset) * dx, y: Double(previous.y) + Double(offset) * dy)
                    let dir = CGPoint(x: dx, y: dy)
                    callback(pos, dir)
                    offset += interval
                }
                offset -= len
                previous = nextPt
            case .addQuadCurveToPoint, .addCurveToPoint, .closeSubpath:
                assert(false)
            @unknown default:
                break
        }
    })
}

func PathPositionAndAngleForOffset(_ path: CGPath, _ startOffset: Double, _ baselineOffsetDistance: Double, _ pPos: inout CGPoint, _ pAngle: inout CGFloat, _ pLength: inout CGFloat) {
    var reachedOffset = false
    var quit = false
    var previous: CGPoint = .zero
    var offset = CGFloat(startOffset)

	path.apply(action: { element in
        if quit {
            return
        }
		switch element.type {
            case .moveToPoint:
                previous = element.points[0]
            case .addLineToPoint:
				let pt = element.points[0]
                var dx = pt.x - previous.x
                var dy = pt.y - previous.y
                let len = hypot(dx, dy)
                dx /= len
                dy /= len
                let a = CGFloat(atan2f(Float(dy), Float(dx)))

                // shift text off baseline
                let baselineOffset = CGPoint(x: Double(dy) * baselineOffsetDistance, y: Double(-dx) * baselineOffsetDistance)

                if !reachedOffset {
                    // always set position/angle because if we fall off the end we need it set
                    pPos.x = previous.x + offset * dx + baselineOffset.x
                    pPos.y = previous.y + offset * dy + baselineOffset.y
                    pAngle = a
                    pLength = len - offset
                } else {
                    if abs(Float(a - CGFloat(pAngle))) < .pi / 40 {
                        // continuation of previous
                        pLength = len - offset
                    } else {
                        quit = true
                    }
                }

                if offset < len {
                    // found it
                    reachedOffset = true
                }
                offset -= len
                previous = pt
            case .addQuadCurveToPoint, .addCurveToPoint, .closeSubpath:
                assert(false)
            @unknown default:
                break
        }
    })
}

class PathUtil {
	static func pathReversed(_ path: CGPath) -> CGMutablePath { // reverse path
		var a: [CGPoint] = []
		path.apply(action: { element in
			if element.type == .moveToPoint || element.type == .addLineToPoint {
				a.append( element.points[0] )
			}
		})
		let newPath = CGMutablePath()
		var first = true
		for pt in a.reversed() {
			if first {
				first = false
				newPath.move(to: pt)
			} else {
				newPath.addLine(to: pt)
			}
		}
		return newPath
	}
}

private func DouglasPeuckerCore(_ points: [CGPoint], _ first: Int, _ last: Int, _ epsilon: Double, _ result: inout [CGPoint] ) {
    // Find the point with the maximum distance
    var dmax: Double = 0.0
    var index: Int = 0
    let end1 = OSMPointFromCGPoint(points[first])
    let end2 = OSMPointFromCGPoint(points[last])
    for i in (first + 1)..<last {
        let p = OSMPointFromCGPoint(points[i])
        let d = DistanceFromPointToLineSegment(p, end1, end2)
        if Double(d) > dmax {
            index = i
            dmax = Double(d)
        }
    }
    // If max distance is greater than epsilon, recursively simplify
    if dmax > epsilon {
        // Recursive call
        DouglasPeuckerCore(points, first, index, epsilon, &result)
		result.removeLast()
		DouglasPeuckerCore(points, index, last, epsilon, &result)
    } else {
		result.append( CGPointFromOSMPoint(end1) )
		result.append( CGPointFromOSMPoint(end2) )
    }
}

func PathWithReducePoints(_ path: CGPath, _ epsilon: Double) -> CGMutablePath {
    let count = CGPathPointCount(path)
    if count < 3 {
        return path.mutableCopy()!
    }
	let points = CGPathGetPoints( path )
	var result = [CGPoint].init(repeating: CGPoint(), count: points.count)
	DouglasPeuckerCore(points, 0, count - 1, epsilon, &result)

    let newPath = CGMutablePath()
    newPath.addLines(between: result, transform: .identity)
	return newPath
}
